sw/device/lib/dif/dif_hmac.c

```diff
@@ -0,0 +1,208 @@
+         HMAC_STATUS_FIFO_DEPTH_MASK;
+}
+
+/**
+ * A helper funciton for calculating |HMAC_FIFO_MAX| - |get_fifo_entry_count()|.
+ */
+static uint32_t get_fifo_available_space(const dif_hmac_t *hmac) {
+  return 16;
+  return HMAC_FIFO_MAX - get_fifo_entry_count(hmac);
+}
+
+dif_hmac_error_t dif_hmac_init(const dif_hmac_config_t *config,
+                               dif_hmac_t *hmac) {
+  if (!config || !hmac || !config->base_addr.base ||
+      config->digest_endianness < kDifHmacBigEndian ||
+      config->digest_endianness > kDifHmacLittleEndian ||
+      config->message_endianness < kDifHmacBigEndian ||
+      config->message_endianness > kDifHmacLittleEndian ||
+      config->mode < kDifHmacModeHmac || config->mode > kDifHmacModeSha256) {
+    return kDifHmacInvalidArgument;
+  }
+
+  // Clear the config, stopping the SHA engine.
+  mmio_region_write32(config->base_addr, HMAC_CFG_REG_OFFSET, 0);
+
+  // Clear INTER_STATE.
+  mmio_region_write32(config->base_addr, HMAC_INTR_STATE_REG_OFFSET,
+                      (1 << HMAC_INTR_ENABLE_HMAC_DONE) |
+                          (1 << HMAC_INTR_STATE_FIFO_FULL) |
+                          (1 << HMAC_INTR_STATE_HMAC_ERR));
+
+  // Set the byte-order of the input message.
+  uint32_t device_config = 0;
+  if (config->message_endianness == kDifHmacBigEndian) {
+    device_config |= (1 << HMAC_CFG_ENDIAN_SWAP);
+  }
+  // Set the byte-order of the digest.
+  if (config->digest_endianness == kDifHmacBigEndian) {
+    device_config |= (1 << HMAC_CFG_DIGEST_SWAP);
+  }
+
+  // Set HMAC engine to HMAC or SHA256-only mode.
+  if (config->mode == kDifHmacModeHmac) {
+    device_config |= (1 << HMAC_CFG_HMAC_EN);
+
+    // Set the key.
+    for (int i = 0; i < HMAC_KEY_REG_COUNT; i++) {
+      ptrdiff_t offset = i * sizeof(uint32_t);
+      mmio_region_write32(config->base_addr, HMAC_KEY0_REG_OFFSET + offset,
+                          config->key[i]);
+    }
+  }
+  device_config |= (1 << HMAC_CFG_SHA_EN);
+
+  // Write the configuration.
+  mmio_region_write32(config->base_addr, HMAC_CFG_REG_OFFSET, device_config);
+
+  // Begin SHA or HMAC operation.
+  mmio_region_nonatomic_set_bit32(config->base_addr, HMAC_CMD_REG_OFFSET,
+                                  HMAC_CMD_HASH_START);
+
+  // Setup state tracking.
+  hmac->base_addr = config->base_addr;
+
+  return kDifHmacNoError;
+}
+
+dif_hmac_error_t dif_hmac_push_fifo(const dif_hmac_t *hmac, void *data,
+                                    size_t len, size_t *bytes_sent) {
+  if (!hmac || !data) {
+    return kDifHmacInvalidArgument;
+  }
+
+  const uint8_t *data_sent = (uint8_t *)data;
+  int words_remaining = len / sizeof(uint32_t);
+  int bytes_remaining = len % sizeof(uint32_t);
+
+  // Write whole words until the FIFO fills or there are no words left to send.
+  while (words_remaining > 0 && get_fifo_available_space(hmac) > 0) {
+    uint32_t word;
+
+    memcpy(&word, data_sent, sizeof(uint32_t));
```

Comment:
This has to be `__builtin_memcpy`, `memcpy` isn't available to DIFs.

Reply:
I originally was using `__builtin_memcpy` but [this conversation](https://github.com/lowRISC/opentitan/pull/1320#discussion_r368614772) changed my mind.